frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,372 @@
+      identityContractID: string // action creator identityContractID
+    }),
+    create () {
+      const { username, identityContractID } = sbp('state/vuex/state').loggedIn
+      return {
+        createdDate: new Date().toISOString(),
+        username,
+        identityContractID
+      }
+    }
+  },
+  state (contractID) {
+    return sbp('state/vuex/state')[contractID]
+  },
+  getters: {
+    currentChatRoomState (state) {
+      return state
+    },
+    chatRoomSettings (state, getters) {
+      return getters.currentChatRoomState.settings || {}
+    },
+    chatRoomAttributes (state, getters) {
+      return getters.currentChatRoomState.attributes || {}
+    },
+    chatRoomUsers (state, getters) {
+      return getters.currentChatRoomState.users || {}
+    },
+    chatRoomLatestMessages (state, getters) {
+      const messages = getters.currentChatRoomState.messages || []
+      return getLatestMessages({
+        count: getters.chatRoomSettings.messagesPerPage,
+        messages
+      })
+    }
+  },
+  actions: {
+    // This is the constructor of Chat contract
+    'gi.contracts/chatroom': {
+      validate: chatRoomType,
+      process ({ meta, data }, { state }) {
+        const initialState = merge({
+          settings: {
+            messagesPerPage: CHATROOM_MESSAGES_PER_PAGE,
+            maxNameLetters: CHATROOM_NAME_LIMITS_IN_CHARS,
+            maxDescriptionLetters: CHATROOM_DESCRIPTION_LIMITS_IN_CHARS
+          },
+          users: {},
+          messages: []
+        }, {
+          attributes: {
+            ...data,
+            creator: meta.username,
+            deletedDate: null,
+            archievedDate: null
+          }
+        })
+        for (const key in initialState) {
+          Vue.set(state, key, initialState[key])
+        }
+      },
+      sideEffect ({ contractID, meta }) {
+        const rootState = sbp('state/vuex/state')
+        if (meta.username === rootState.loggedIn.username) {
+          sbp('state/vuex/commit', 'setCurrentChatRoomId', { chatRoomId: contractID })
+        }
+      }
+    },
+    'gi.contracts/chatroom/join': {
+      validate: objectMaybeOf({
+        username: string,
+        referer: string
+      }),
+      process ({ data, meta, hash }, { state }) {
+        const { username, referer } = data
+
+        const notificationType = username === referer ? MESSAGE_NOTIFICATIONS.JOIN_MEMBER : MESSAGE_NOTIFICATIONS.ADD_MEMBER
+        const notificationData = createNotificationData(
+          notificationType,
+          notificationType === MESSAGE_NOTIFICATIONS.ADD_MEMBER ? { username } : {}
+        )
+        const newMessage = createMessage({ meta, hash, data: notificationData, state })
+        Vue.set(state.messages, [state.messages.length], newMessage)
+
+        if (state.users[username] && !state.users[username].departedDate) {
+          console.log(`chatroom Join: ${username} is already joined the chatroom #${state.name}`)
+          return
+        }
```

Comment:
- Error checking should always be done at the very beginning of the process function
- If something should never be possible through the UI, and is likely a hacking attempt, an exception should be thrown using the appropriate type of exception (see example below)
- If it's an error, `console.error` should be used

Example from `group.js`:

```js
    'gi.contracts/group/proposalVote': {
      validate: objectOf({
        proposalHash: string,
        vote: string,
        passPayload: optional(unionOf(object, string)) // TODO: this, somehow we need to send an OP_KEY_ADD GIMessage to add a generated once-only writeonly message public key to the contract, and (encrypted) include the corresponding invite link, also, we need all clients to verify that this message/operation was valid to prevent a hacked client from adding arbitrary OP_KEY_ADD messages, and automatically ban anyone generating such messages
      }),
      process (message, { state }) {
        const { data, hash, meta } = message
        const proposal = state.proposals[data.proposalHash]
        if (!proposal) {
          // https://github.com/okTurtles/group-income-simple/issues/602
          console.error(`proposalVote: no proposal for ${data.proposalHash}!`, { data, meta, hash })
          throw new Errors.GIErrorIgnoreAndBanIfGroup('proposalVote without existing proposal')
        }
```